### PR TITLE
8336245: AArch64: remove extra register copy when converting from long to pointer

### PR DESCRIPTION
**Pull request by fg1417**
In the cases like:
```
  UNSAFE.putLong(address + off1 + 1030, lseed);
  UNSAFE.putLong(address + 1023, lseed);
  UNSAFE.putLong(address + off2 + 1001, lseed);
```

Unsafe intrinsifies direct memory access using a long as the base address, generating a `CastX2P` node converting long to pointer in C2. Then we get optoassembly code like:
```
  ldr  R10, [R15, #120]    # int ! Field: address
  ldr  R11, [R16, #136]    # int ! Field: off1
  ldr  R12, [R16, #144]    # int ! Field: off2
  add  R11, R11, R10
  mov R11, R11    # long -> ptr
  add  R12, R12, R10
  mov R10, R10    # long -> ptr
  add R11, R11, #1030    # ptr
  str  R17, [R11]    # int
  add R10, R10, #1023    # ptr
  str  R17, [R10]    # int
  mov R10, R12    # long -> ptr
  add R10, R10, #1001    # ptr
  str  R17, [R10]    # int
```

In aarch64, the conversion from long to pointer could be a nop but C2 doesn't know it. On the existing code, we do nothing for `mov dst src` only when `dst` == `src` [1], then we have assembly:
```
  ldr    x10, [x15,#120]
  ldp    x11, x12, [x16,#136]
  add    x11, x11, x10
  add    x12, x12, x10
  add    x11, x11, #0x406
  str    x17, [x11]
  add    x10, x10, #0x3ff
  str    x17, [x10]
  mov    x10, x12  <--- extra register copy
  add    x10, x10, #0x3e9
  str    x17, [x10]
```

There is still one extra register copy, which we're trying to remove in this patch.

This patch folds `CastX2P` into memory operands by introducing `indirectX2P` and `indOffX2P`. We also create a new opclass `iRegPorL2P` to remove extra copies from `CastX2P` in pointer addition.

Tier 1~3 passed on aarch64. No obvious change in size of libjvm.so

[1] https://github.com/openjdk/jdk/blob/5c612c230b0a852aed5fd36e58b82ebf2e1838af/src/hotspot/cpu/aarch64/aarch64.ad#L7906

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8336245](https://bugs.openjdk.org/browse/JDK-8336245): AArch64: remove extra register copy when converting from long to pointer (**Enhancement** - P4)

### Reviewers
 * [Andrew Haley](https://openjdk.org/census#aph) (@theRealAph - **Reviewer**)
 * [Andrew Dinn](https://openjdk.org/census#adinn) (@adinn - **Reviewer**)

### Contributors
 * Andrew Haley `<aph@openjdk.org>`

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/20157/head:pull/20157` \
`$ git checkout pull/20157`

Update a local copy of the PR: \
`$ git checkout pull/20157` \
`$ git pull https://git.openjdk.org/jdk.git pull/20157/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 20157`

View PR using the GUI difftool: \
`$ git pr show -t 20157`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/20157.diff">https://git.openjdk.org/jdk/pull/20157.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/20157#issuecomment-2225635287)

### CODE REVIEW

File: src/hotspot/share/opto/machnode.cpp
Comment (fg1417) [inline src/hotspot/share/opto/machnode.cpp:400]: After applying the operand "IndirectX2P", we may have some patterns like:
```
str val, [CastX2P base]
```
The code path here will resolve the `base`, which is actually a `intptr`, not a `ptr`, and the offset is `0`.

I guess the code here was intended to support `[base, offset]`, where base can be a `intptr` but offset can not be `0`. I'm not sure why there is such a limitation that offset can not be `0`, maybe for some old machines?

I don't think the limitation is applied to aarch64 machines now. So I unblock it for aarch64.

**Comment by fg1417** [comment]
https://github.com/openjdk/jdk/pull/20159 is also to fix the same issue. Please feel free to review the draft PR. Thanks.

File: src/hotspot/share/opto/machnode.cpp
Comment (theRealAph) [reply src/hotspot/share/opto/machnode.cpp:400]: I think it's the other way around. Isn't this code saying that if the address is an intptr + a nonzero offset, then the returned type is bottom, ie nothing? What effect does this change have?

Comment (theRealAph) [reply src/hotspot/share/opto/machnode.cpp:400]: I learn something every day, I guess. It's been a long while since I looked, but I expected "pointer to anything" to be TOP, not BOTTOM. Thinking some more, a `TypeRawPtr` must be casted to a usable physical type in order to use it, so "pointer to nothing" makes more sense than "pointer to anything".

**Review by theRealAph — APPROVED** [review]

**Comment by theRealAph** [comment]
This will need quite a lot of testing, perhaps higher tiers and jcstress.

**Comment by fg1417** [comment]
>> #20159 is also to fix the same issue. Please feel free to review the draft PR. Thanks.

> This will need quite a lot of testing, perhaps higher tiers and jcstress. You can test these two PRs together.

Thanks for approval, @theRealAph. I'll test jcstress on my local.

Could you help review and test these two PRs with higher tiers please? @TobiHartmann @vnkozlov Thanks.

**Comment by TobiHartmann** [comment]
Sure, I'll run this through our testing and report back once it passed.

**Comment by TobiHartmann** [comment]
All tests passed.

**Comment by fg1417** [comment]
> All tests passed.

@TobiHartmann thanks for your testing!

Jcstress also passed.

**Comment by fg1417** [comment]
Can I have a second review please :-)

File: src/hotspot/cpu/aarch64/aarch64.ad
Comment (adinn) [inline src/hotspot/cpu/aarch64/aarch64.ad:4235]: Why is this using hard wired constants rather than using Address::offset_ok_for_immed?

Also, why is the constant value 65520?

Comment (adinn) [reply src/hotspot/cpu/aarch64/aarch64.ad:4235]: Yes, I realise that this is 16 less than 65536. However, there are two things I don't follow.

In the original code immLoffset was only used to define indOffLN i.e. a long offset used with a narrow pointer. The use of Address::offset_ok_for_immed(n->get_long(), 0) in the predicate limited narrow pointer offsets to -256 <= offset <= (2^12 - 1). With this change the top end of the range is now (2^12 - 1) << 4. I am wondering why that is appropriate?

The change allows immLOffset to be used in the definition of indOffX2P. I am not clear why indOffX2P is not just defined using the existing operand immLoffset16 which has as its predicate Address::offset_ok_for_immed(n->get_long(), 4). The only difference I can see is that the alternative predicate used here will accept a positive offset that is not 16 byte aligned. Is that the intention of the redefinition? Again, why is that appropriate?

Comment (adinn) [reply src/hotspot/cpu/aarch64/aarch64.ad:4235]: Ok, I see. The use of immLoffset as currently defined is actually correct for narrow oops and, indeed, for all other address base types. It allows for all possible offsets that might fit into a load an immediate slot. Whether we can legitimately encode the operand offset as an immediate or need instead to use an auxiliary add does not actually depend on the type of the address base but on the size of the datum fetched by the indirect load that consumes the operand. So, an indirect operand with offset 4098 would be too big to encode in an ldrb, fine to encode in an ldrh and invalid for encoding in an ldrw or ldrx because it is not suitably aligned.

That does imply we should get rid of the other (now redundant) immLoffset<n> operands. However, we can do that in a follow-up patch because it is not what this fix is addressing

**Review by adinn — APPROVED** [review]

**Comment by fg1417** [comment]
@theRealAph @adinn Thanks for your reviews!

I'll integrate it.

/integrate

**Pull request closed without merging**
